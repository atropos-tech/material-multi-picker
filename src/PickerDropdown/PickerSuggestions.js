import React from "react";
import { MenuItem, withStyles } from "@material-ui/core";
import { LOADING, NOT_ENOUGH_CHARACTERS, isError, suggestionsPropType } from "../utils";
import { string, func, number, any, object } from "prop-types";
import DefaultError from "./DefaultError";
import DefaultSuggestion from "./DefaultSuggestion";
import DefaultEmptyMessage from "./DefaultEmptyMessage";
import DefaultLoadingMessage from "./DefaultLoadingMessage";
import DefaultMoreCharactersMessage from "./DefaultMoreCharactersMessage";

const HIGHLIGHT_GREY_SHADE = 300;

const styles = theme => ({
    highlighted: {
        backgroundColor: theme.palette.grey[HIGHLIGHT_GREY_SHADE]
    }
});

function PickerSuggestions(
    { suggestions, getItemProps, highlightedIndex, itemToString, inputValue, SuggestionComponent = DefaultSuggestion, ErrorComponent = DefaultError, classes }
) {
    if ( isError(suggestions) ) {
        return (<ErrorComponent error={ suggestions } inputValue={ inputValue } />);
    }
    if ( suggestions === NOT_ENOUGH_CHARACTERS ) {
        return <DefaultMoreCharactersMessage />;
    }
    if ( suggestions === LOADING ) {
        return <DefaultLoadingMessage inputValue={ inputValue } />;
    }
    if ( Array.isArray(suggestions) ) {
        if ( suggestions.length ) {
            return (
                <>
                    {
                        suggestions.map((item, index) => {
                            const itemId = itemToString(item);
                            const isHighlighted = highlightedIndex === index;

                            const menuItemProps = getItemProps({
                                index,
                                item,
                                role: "menuitem",
                                className: isHighlighted ? classes.highlighted : "",
                                style: {
                                    padding: 0,
                                    height: "auto"
                                },
                            });
                            return (
                                <MenuItem className="suggestion" key={ itemId } {...menuItemProps}>
                                    <SuggestionComponent itemId={ itemId } item={ item } isHighlighted={ isHighlighted } inputValue={ inputValue } />
                                </MenuItem>
                            );
                        })
                    }
                </>
            );
        }
        if ( inputValue.length ) {
            return <DefaultEmptyMessage inputValue={ inputValue } />;
        } else {
            return false;
        }
    }

    console.error(suggestions);
    throw new Error("should never happen!");
}

PickerSuggestions.propTypes = {
    suggestions: suggestionsPropType,
    getItemProps: func.isRequired,
    highlightedIndex: number,
    itemToString: func.isRequired,
    inputValue: string.isRequired,
    SuggestionComponent: any,
    ErrorComponent: any,
    classes: object
};

export default withStyles(styles)(PickerSuggestions);
