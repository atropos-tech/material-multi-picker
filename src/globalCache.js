const caches = {};

function createGlobalCache() {
    const cachedData = new Map();
    const updateListeners = [];

    function notifyUpdate() {
        updateListeners.forEach(listener => {
            try {
                listener();
            } catch (error) {
                console.error(error);
            }
        });
    }

    function removeListener(listener) {
        const listenerIndex = updateListeners.indexOf(listener);
        if ( listenerIndex >= 0 ) {
            updateListeners.splice(listenerIndex, 1);
        }
    }

    return {
        getValue(key) {
            return cachedData.get(key);
        },
        setValue(key, value) {
            cachedData.set(key, value);
            notifyUpdate();
        },
        clearAll() {
            cachedData.clear();
            notifyUpdate();
        },
        subscribeToUpdates(updateListener) {
            updateListeners.push(updateListener);
            return () => removeListener(updateListener);
        },
        getListenerCount() {
            return updateListeners.length;
        }
    };
}

export function getGlobalCache(id) {
    if (!caches[id]) {
        caches[id] = createGlobalCache();
    }
    return caches[id];
}

export function resetAllCaches() {
    Object.keys(caches).forEach(key => {
        delete caches[key];
    });
}


