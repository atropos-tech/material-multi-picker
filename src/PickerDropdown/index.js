import React from "react";
import { createPortal } from "react-dom";
import { Paper } from "@material-ui/core";
import PickerSuggestions from "./PickerSuggestions";
import { bool, number, string, array, func } from "prop-types";
import debounceRender from "react-debounce-render";
import useCachedSuggestions from "./useCachedSuggestions";

const DELAYED_RENDER_MILLISECONDS = 30;

const DROPDOWN_STYLE = {
    position: "absolute",
    zIndex: 1400,
    width: "100%",
    overflowY: "auto",
    padding: 0,
    margin: 0
};

function getRelativeDropdownPositionStyle(anchorElement) {
    if ( anchorElement && anchorElement.offsetParent ) {
        const { offsetParent } = anchorElement;
        return {
            left: offsetParent.offsetLeft,
            right: offsetParent.offsetLeft + offsetParent.offsetWidth,
            top: offsetParent.offsetTop + offsetParent.offsetHeight
        };
    }
    return {};
}

//adapted from https://plainjs.com/javascript/styles/get-the-position-of-an-element-relative-to-the-document-24/
function getDocumentBoundingRect(element) {
    const viewportRectangle = element.getBoundingClientRect();
    return {
        top: viewportRectangle.top + window.pageYOffset,
        left: viewportRectangle.left + window.pageXOffset,
        bottom: viewportRectangle.bottom + window.pageYOffset,
        right: viewportRectangle.left + window.pageXOffset,
        width: viewportRectangle.width,
        height: viewportRectangle.height
    };
}

function getViewportDropdownPositionStyle(anchorElement) {
    if ( anchorElement && anchorElement.offsetParent ) {
        const { offsetParent } = anchorElement;
        const anchorRectangle = getDocumentBoundingRect(offsetParent);

        return {
            left: anchorRectangle.left,
            width: anchorRectangle.width,
            top: anchorRectangle.bottom
        };
    }
    return {};
}

function getDropdownStyle(anchorElement, relativeToOffsetParent, maxHeight) {
    const dropdownPositionStyle = relativeToOffsetParent ? getRelativeDropdownPositionStyle(anchorElement) : getViewportDropdownPositionStyle(anchorElement);
    const dropdownMaxHeightStyle = maxHeight ? { maxHeight } : {};
    return {
        ...DROPDOWN_STYLE,
        ...dropdownPositionStyle,
        ...dropdownMaxHeightStyle
    };
}

function useUnpickedSuggestions(isReadyToLoad, inputValue, pickedItems, getSuggestedItems, itemToString, useGlobalCache, fetchDelay) {
    const suggestions = useCachedSuggestions(isReadyToLoad, inputValue, pickedItems, getSuggestedItems, useGlobalCache, fetchDelay);
    if (Array.isArray(suggestions) && Array.isArray(pickedItems)) {
        const pickedItemIds = pickedItems.map(item => itemToString(item));
        const isUnpicked = suggestion => !pickedItemIds.includes(itemToString(suggestion));
        return suggestions.filter(isUnpicked);
    }
    return suggestions;
}

function Dropdown({ isOpen, suggestions, anchorElement, maxHeight, disablePortals, ...otherProps }) {
    if ( isOpen && suggestions ) {
        const dropdownStyle = getDropdownStyle(anchorElement, disablePortals, maxHeight);
        const dropdown = (
            <Paper role="menu" component="ul" square style={ dropdownStyle }>
                <PickerSuggestions suggestions={ suggestions } { ...otherProps } />
            </Paper>
        );
        return disablePortals ? dropdown : createPortal(dropdown, document.body);
    }
    return false;
}

Dropdown.propTypes = {
    isOpen: bool,
    maxHeight: number
};

const DebouncedDropdown = debounceRender(Dropdown, DELAYED_RENDER_MILLISECONDS);

function PickerDropdown(props) {
    const { isOpen, inputValue, pickedItems, getSuggestedItems, itemToString, useGlobalCache, fetchDelay } = props;
    const suggestions = useUnpickedSuggestions(isOpen, inputValue, pickedItems, getSuggestedItems, itemToString, useGlobalCache, fetchDelay);

    return <DebouncedDropdown {...props} suggestions={ suggestions } />;
}

PickerDropdown.propTypes = {
    isOpen: bool,
    inputValue: string,
    pickedItems: array,
    getSuggestedItems: func.isRequired,
    itemToString: func.isRequired,
    useGlobalCache: string,
    fetchDelay: number,
    disablePortals: bool
};

export default PickerDropdown;
