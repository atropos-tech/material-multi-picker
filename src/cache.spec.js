/* eslint-env jest */
/* eslint-disable no-magic-numbers */

import React from "react";
import { mount } from "enzyme";
import { resetIdCounter } from "downshift";
import MultiPicker from "./index";
import { Paper } from "@material-ui/core";
import JssProvider from "react-jss/lib/JssProvider";
import { LOADING } from "./utils";
import { getGlobalCache, resetAllCaches } from "./PickerDropdown/globalCache";

// workaround for non-stable classnames generated in JSS
// https://github.com/mui-org/material-ui/issues/9492#issuecomment-368205258
//
// forces deterministic classnames that may clash in a wider app (but are safe in unit tests)
//
// Should be addressed in future material-ui release, see
//   https://github.com/mui-org/material-ui/issues/14357
function mountStable(reactElement, ...other) {
    const generateClassName = (rule, styleSheet) => `${styleSheet.options.classNamePrefix}-${rule.key}`;
    const elementWithStableClassNames = (
        <JssProvider generateClassName={ generateClassName }>
            { reactElement }
        </JssProvider>
    );
    return mount(elementWithStableClassNames, ...other);
}

const NOOP = () => { /* do nothing */ };
const SHORT_DELAY_MILLISECONDS = 30;

function delay(delayInMilliseconds = SHORT_DELAY_MILLISECONDS) {
    return new Promise(resolve => {
        setTimeout(resolve, delayInMilliseconds);
    });
}

async function changeInputValueAndUpdate(wrapper, newInputValue) {
    wrapper.find("input").simulate("change", { target: { value: newInputValue}});
    await delay(100);
    wrapper.update();
}

describe("MultiPicker component", () => {

    // avoid non-deterministic id creation for downshift subcomponents
    // breaking snapshot testing
    beforeEach(() => {
        resetIdCounter();
        resetAllCaches();
    });

    it("uses global cache if configured", async () => {
        expect.assertions(5);

        const cache = getGlobalCache("test");

        jest.spyOn(cache, "setValue");

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => ["some suggestion"],
            useGlobalCache: "test",
            disablePortals: true
        };
        const wrapper = mountStable(<MultiPicker {...props }/>);

        await changeInputValueAndUpdate(wrapper, "some text");

        expect(cache.setValue).toHaveBeenCalledWith("some text", LOADING);
        expect(cache.setValue).toHaveBeenCalledWith("some text", ["some suggestion"]);

        expect(cache.getValue("some text")).toEqual(["some suggestion"]);
        expect(wrapper.find(Paper)).toHaveText("some suggestion");

        expect(wrapper.find(Paper).getDOMNode()).toMatchSnapshot();
    });

    it("subscribes and unsubscribes from global cache if configured", () => {
        expect.assertions(3);

        const cache = getGlobalCache("test");

        jest.spyOn(cache, "subscribeToUpdates");

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => ["some suggestion"],
            useGlobalCache: "test",
            disablePortals: true
        };
        const wrapper = mountStable(<MultiPicker {...props }/>);

        expect(cache.subscribeToUpdates).toHaveBeenCalled();
        expect(cache.getListenerCount()).toEqual(1);

        wrapper.unmount();

        expect(cache.getListenerCount()).toEqual(0);
    });


    it("does not touch global cache if not configured", async () => {
        expect.assertions(1);

        const cache = getGlobalCache("test");

        jest.spyOn(cache, "subscribeToUpdates");

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => ["some suggestion"],
            disablePortals: true
        };
        const wrapper = mountStable(<MultiPicker {...props }/>);

        await changeInputValueAndUpdate(wrapper, "some text");
        expect(cache.subscribeToUpdates).not.toHaveBeenCalled();
    });

    it("shares global cache between instances", async () => {
        expect.assertions(4);

        const cache = getGlobalCache("test");

        jest.spyOn(cache, "subscribeToUpdates");

        const props1 = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: jest.fn(() => ["some suggestion"]),
            useGlobalCache: "test",
            disablePortals: true
        };
        const props2 = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: jest.fn(() => ["some suggestion"]),
            useGlobalCache: "test"
        };
        const wrapper1 = mountStable(<MultiPicker {...props1 }/>);
        const wrapper2 = mountStable(<MultiPicker {...props2 }/>);

        await changeInputValueAndUpdate(wrapper1, "some text");
        expect(props1.getSuggestedItems).toHaveBeenCalled();
        expect(cache.getValue("some text")).toEqual(["some suggestion"]);

        await changeInputValueAndUpdate(wrapper2, "some text");
        expect(props2.getSuggestedItems).not.toHaveBeenCalled();
        expect(wrapper2.find(Paper)).toHaveText("some suggestion");
    });

});
