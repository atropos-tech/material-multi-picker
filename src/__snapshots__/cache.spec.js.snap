// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiPicker component uses global cache if configured 1`] = `
<ul
  class="MuiPaper-root MuiPaper-elevation2"
  role="menu"
  style="position: absolute; z-index: 1400; width: 100%; overflow-y: auto; padding: 0px; margin: 0px;"
>
  <li
    aria-selected="false"
    class="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root MuiMenuItem-gutters"
    id="downshift-0-item-0"
    role="menuitem"
    style="padding: 0px; height: auto;"
    tabindex="-1"
  >
    <p
      class="MuiTypography-root MuiTypography-body1"
      style="padding: 11px 16px;"
    >
      some suggestion
    </p>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</ul>
`;
